test/safe.test.ts

```diff
@@ -0,0 +1,197 @@
+import { TransactionResponse } from "@ethersproject/abstract-provider";
+import { expect, use } from "chai";
+import chaiAsPromised from "chai-as-promised";
+import { BigNumber, Contract, utils, Wallet } from "ethers";
+import hre, { ethers, waffle } from "hardhat";
+
+import { gnosisSafeAt } from "../src/tasks/ts/safe";
+import {
+  createdProxies,
+  prepareDeterministicSafeWithOwners,
+  prepareSafeWithOwners,
+  getFallbackHandler,
+} from "../src/ts/lib/safe";
+
+import { GnosisSafeManager } from "./safe";
+
+use(chaiAsPromised);
+
+const [deployer, creator, ...owners] = waffle.provider.getWallets();
+
+async function safeAddress(
+  response: TransactionResponse,
+  factory: string,
+): Promise<string> {
+  const proxies = await createdProxies(response, factory);
+  expect(proxies).to.have.length(1);
+  return proxies[0];
+}
+
+type DeployFunction = (
+  owners: Wallet[],
+  threshold: number,
+) => Promise<[TransactionResponse, GnosisSafeManager]>;
+
+function testSafe(deploy: DeployFunction) {
```

Comment:
nit: Maybe testSafeSetup() would be a better name.

But I really like how you were writing this test!